### DISCUSSION TITLE
Repeater Move/Delete Hooks

### DISCUSSION THREAD

**Discussion by tangjh** [filamentphp/filament#1222] [Help]
Is there any hooks / callback that can be latched on when user clicks on move up/down or delete button in a Repeater?
eg.  beforeDelete, afterDelete, beforeMoveUp, afterMoveUp, ...

I tried reactive, afterStateUpdated on the Repeater but it isn't called.

**Comment by titantwentyone** [comment]
You can register a listener which will be invoked directly after the logic for any of the actions:

```
Repeater::make('repeater')
    ->schema([ // ])
    ->registerListeners([
        'repeater::createItem' => function(Repeater $component, string $statePath): void {
            //your logic here
        }
    ])
```
Have a look [here](https://github.com/laravel-filament/filament/blob/5ee3d746892aa8b5c36a4ea0fdc14f606492881f/packages/forms/src/Components/Repeater.php#L39) to see how they are used within the Repeater component itself.

These aren't documented so not sure if okay to use. I happened across them as the same functionality is used in the Builder component.

**Reply by danharrin** [reply]
No guarantees that there aren't breaking changes for the internal event names for our components, but the event system itself is documented so isnt going anywhere.

https://filamentadmin.com/docs/2.x/forms/advanced#using-form-events

**Reply by titantwentyone** [reply]
Sorry! It is documented! Totally forgot. As an aside to this, is there a mechanism to invoke code before an event occurs? At the moment, the code executes after the components own event handlers. e.g.  i want to invoke logic before a repeater deletes an item. The only way at the moment, I think, is to clear the all listeners and rebuild them.

**Reply by danharrin** [reply]
Not at the moment, you'd need to create a custom Repeater that overrides `setUp()` to register your event listeners before calling `parent::setUp()`

**Comment by tangjh** [comment]
In case anybody interested, i did a sed to update the project's vendor Repeater.php source whenever update occurs (quick and dirty as i ran out of time):

            'repeater::deleteItem' => [
                function (Repeater $component, string $statePath, string $uuidToDelete): void {
                    if ($component->isDisabled()) {
                        return;
                    }

                    if ($statePath !== $component->getStatePath()) {
                        return;
                    }

                    $items = $component->getState();

                    unset($items[$uuidToDelete]);

                    // workaround for after delete event
                    $livewire = $component->getLivewire();
                    $livewire->dispatchFormEvent('repeater::afterDeleteItem', $livewire);
                },
            ],

....
Then in the repeater component:
...
                                Forms\Components\HasManyRepeater::make('your_rep')
                                    ->...
                                    ->registerListeners([
                                        'repeater::afterDeleteItem' => [
                                            function (Forms\Components\Component $component, $livewire): void {
                                                // do whatever you need
                                            },
                                        ],
                                    ])
                                    


You can apply same principle for beforeDeleteItem, beforeCreateItem, afterCreateItem, ....
